emission/analysis/modelling/trip_model/greedy_similarity_binning.py

```diff
@@ -0,0 +1,252 @@
+import logging
+from tokenize import group
+from typing import Dict, List, Optional, Tuple
+
+import emission.analysis.modelling.similarity.similarity_metric_type as eamssmt
+import emission.analysis.modelling.similarity.similarity_metric as eamss
+import emission.analysis.modelling.tour_model.label_processing as lp
+import emission.analysis.modelling.trip_model.trip_model as eamuu
+import emission.analysis.modelling.trip_model.util as util
+import emission.analysis.modelling.trip_model.config as eamtc
+import emission.core.wrapper.confirmedtrip as ecwc
+import pandas as pd
+
+
+class GreedySimilarityBinning(eamuu.TripModel):
+
+    is_incremental: bool = False  # overwritten during __init__
+
+    def __init__(self, config=None):
+        """
+        instantiate a clustering model for a user.
+
+        replaces the original similarity class
+        [https://github.com/e-mission/e-mission-server/blob/5b9e608154de15e32df4f70a07a5b95477e7dbf5/emission/analysis/modelling/tour_model/similarity.py#L67]
+
+        this technique employs a greedy similarity heuristic to associate
+        trips with collections of probabilistic class labels. in pseudocode:
+        
+        # fit
+        for each bin_id, bin in bins:
+            for each bin_trip in bin.trips:
+                if similar(trip, bin_trip):
+                    append trip to bin.trips
+
+        # prediction
+        for each bin_id, bin in bins:
+            for each bin_trip in bin.trips:
+                if similar(trip, bin_trip):
+                    return bin.predictions: List[Prediction]
+
+        the number of predictions is not assumed to be the number of features.
+
+        the original similarity class (link above) used a nested List data 
+        structure to capture the notion of binning. this was then copied into
+        a Dict when the model needed to be saved. the same technique can be 
+        written to work directly on nested Dicts with no loss in performance. 
+        the data takes the form:
+        {
+            bin_id: {
+                "features": [
+                    [f1, f2, .., fn],
+                    ...
+                ],
+                "labels": [
+                    { label1: value1, ... }
+                ],
+                "predictions": [
+                    { "labels": { label1: value1, ... }, 'p': p_val }
+                ]
+            }
```

Comment:
> why do we have "features" instead of "trips"

1. since this object is also the one we will store in the database, it is smaller if it only includes the features we require for classification
2. trips are domain-specific, i thought for most folks just working on classification they would prefer working in the domain of real number vectors
3. this way, we extract trip features at most once

> why do we save the labels and predictions separately? I double-checked the current file-based storage, and

labels are the input and are extracted when we extract the features. predictions are tacked on once we finish fitting the model. if optimizing memory/storage size is important, this could be re-written so that we remove "labels" once we finish creating "predictions". but
1. when re-running an incremental model, we can apply the same code that trains from "labels" if we still have them (as opposed to reverse-engineering the labels from the previous predictions)
2. to a smaller degree, leaving the labels does help with debugging and explaining the model

Reply:
> it should definitely be explained better in the comments

sure. but, i mean, the same could be said about the previous implementation.